### PR TITLE
casadi: build on darwin

### PR DESCRIPTION
**Pull request by nim65s**
## Description of changes

Follows #331343, and require #333016. When the latter is merged, I'll mark this ready to review.
Pinocchio's `example-py-casadi-quadrotor-ocp` which use casadi + ipopt + mumps from python is also working on darwin now.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/science/math/bonmin/default.nix [hunk 1]
```diff
@@ -27,6 +30,7 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [
+    doxygen
```
Comment (doronbehar) [inline pkgs/development/libraries/science/math/bonmin/default.nix:33]: This change should be in a separate commit IMO.

File: pkgs/development/libraries/science/math/bonmin/default.nix [hunk 2]
```diff
@@ -35,20 +39,34 @@ stdenv.mkDerivation rec {
     bzip2
     cbc
     clp
+    coin-utils
     ipopt
     lapack
     libamplsolver
+    osi
```
Comment (doronbehar) [inline pkgs/development/libraries/science/math/bonmin/default.nix:46]: Same as this change - it'd be nice to explain in a comment what do these new deps add.

File: pkgs/development/libraries/science/math/bonmin/default.nix [hunk 3]
```diff
@@ -35,20 +39,34 @@ stdenv.mkDerivation rec {
...
+
+  # ignore one failing test
+  postPatch = ''
+    substituteInPlace Bonmin/test/Makefile.in --replace-fail \
+      "./unitTest\''$(EXEEXT)" \
+      ""
+  '';
+
+  doCheck = true;
+  checkTarget = "test";
```
Comment (doronbehar) [inline pkgs/development/libraries/science/math/bonmin/default.nix:62]: Also adding the tests should be in a separate commit.

File: pkgs/development/libraries/science/math/bonmin/default.nix [hunk 4]
```diff
@@ -35,20 +39,34 @@ stdenv.mkDerivation rec {
...
+  configureFlagsArray = [
+    "--with-asl-lib=-lipoptamplinterface -lamplsolver"
+  ];
+
+  # ignore one failing test
+  postPatch = ''
+    substituteInPlace Bonmin/test/Makefile.in --replace-fail \
+      "./unitTest\''$(EXEEXT)" \
+      ""
+  '';
```
Comment (doronbehar) [inline pkgs/development/libraries/science/math/bonmin/default.nix:59]: And then I assume either of these is only relevant to Darwin?

File: pkgs/by-name/hp/hpipm/package.nix [hunk 5]
```diff
@@ -29,6 +29,13 @@ stdenv.mkDerivation (finalAttrs: {
     "-DBLASFEO_PATH=${blasfeo}"
   ] ++ lib.optionals (!stdenv.isx86_64) [ "-DTARGET=GENERIC" ];
 
+  postPatch = ''
+    # build shared lib
+    substituteInPlace CMakeLists.txt --replace-fail \
+      "add_library(hpipm " \
+      "add_library(hpipm SHARED "
+  '';
```
Comment (doronbehar) [inline pkgs/by-name/hp/hpipm/package.nix:37]: Hmm it would be nice if upstream would have supported that naturally. Perhaps something a bit more elegant could be done with [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html)?

File: pkgs/by-name/ca/casadi/package.nix [hunk 6]
```diff
@@ -30,7 +30,7 @@
   #sundials,
   superscs,
   spral,
-  swig,
```
Comment (doronbehar) [inline pkgs/by-name/ca/casadi/package.nix:33]: Also this change looks logically separate from the Dawrin build change.

Comment (doronbehar) [reply pkgs/by-name/ca/casadi/package.nix:33]: > it is not: swig 3 generated code contains things in C++ std which were removed in C++17, so it won't build on darwin. But you're totally right: it looks like it is separate, so I'm adding a comment to document that.

Still though, if it builds fine with swig 4 on Linux as well, it can be considered a separate change, as even Linux users will want to know that the latest and greatest swig generated the C++ code that is part of this package.

File: pkgs/by-name/hp/hpipm/package.nix (hunk 5 above)
Comment (nim65s) [reply pkgs/by-name/hp/hpipm/package.nix:37]: Hum, I packaged this looking at the CMakeLists.txt of 0.1.3, but in 0-unstable-2024-07-30 it was greatly improved. BUILD_SHARED_LIBS is already there :),

**Comment by nim65s** [comment]
@doronbehar : thanks a lot for your review !
I think I addressed every point. 

nixpkgs-review new results should be posted soon on aarch64-darwin & x86_64-linux.

File: pkgs/by-name/hp/hpipm/package.nix [hunk 7]
```diff
@@ -26,7 +26,8 @@ stdenv.mkDerivation (finalAttrs: {
   ];
 
   cmakeFlags = [
-    "-DBLASFEO_PATH=${blasfeo}"
+    "-DHPIPM_FIND_BLASFEO=ON"
```
Comment (doronbehar) [inline pkgs/by-name/hp/hpipm/package.nix:29]: Is this flag equivalent to the previous `"-DBLASFEO_PATH=${blasfeo}"` flag? Even if so, it would be nice if this change could be in a separate commit, or explained in the commit message.

**Review by doronbehar — COMMENTED** [review]
Changes look mostly good. BTW this should be marked as ready for review as #333016 is merged.

**Comment by nim65s** [comment]
some tests are failing for now

**Comment by nim65s** [comment]
Result of `nixpkgs-review pr 333314` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>bonmin</li>
    <li>casadi</li>
    <li>crocoddyl</li>
    <li>hpipm</li>
    <li>pinocchio</li>
    <li>python311Packages.casadi</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.pinocchio</li>
    <li>python312Packages.casadi</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.pinocchio</li>
  </ul>
</details>

**Comment by nim65s** [comment]
Result of `nixpkgs-review pr 333314` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>bonmin</li>
    <li>casadi</li>
    <li>crocoddyl</li>
    <li>pinocchio</li>
    <li>python311Packages.casadi</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.pinocchio</li>
    <li>python312Packages.casadi</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.pinocchio</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>hpipm</li>
    <li>ipopt</li>
    <li>mumps</li>
    <li>openturns</li>
    <li>python311Packages.openturns</li>
    <li>python311Packages.pygmo</li>
    <li>python312Packages.openturns</li>
    <li>python312Packages.pygmo</li>
  </ul>
</details>

**Comment by nim65s** [comment]
No, this is still wrong, sorry for the noise. Casadi should build on darwin, that's the point of this PR…

**Comment by nim65s** [comment]
I don't understand why nixpkgs-review does not list casadi on aarch64-darwin, but everything seems fine.

**Comment by wegank** [comment]
> I don't understand why nixpkgs-review does not list casadi on aarch64-darwin, but everything seems fine.

Does `casadi` have `meta.platforms` set?

**Comment by nim65s** [comment]
> Does `casadi` have `meta.platforms` set?

no. But I can find `165/165 Test #165: example-py-casadi-quadrotor-ocp ....................   Passed    7.15 sec` in `nix log .#python3Packages.pinocchio` on aarch64-darwin, so I think this is fine.

**Comment by doronbehar** [comment]
> I don't understand why nixpkgs-review does not list casadi on aarch64-darwin, but everything seems fine.

That's indeed very peculiar.. The ofborg build attempts of the relevant attributes on `x86_64-darwin` were all successful, and for `aarch64-linux` they don't appear as green, and it is not fully clear to me why:

```
error: 1 dependencies of derivation '/nix/store/dwicrqshl3nmw9qc6jxaf0ksrji02smc-casadi-3.6.6.drv' failed to build
error: 1 dependencies of derivation '/nix/store/006w9vz1914vyigmnaj31ks8vhnmml2h-pinocchio-3.1.0.drv' failed to build
error: build of '/nix/store/006w9vz1914vyigmnaj31ks8vhnmml2h-pinocchio-3.1.0.drv', '/nix/store/dwicrqshl3nmw9qc6jxaf0ksrji02smc-casadi-3.6.6.drv' failed
```

Also, `hydra-check casadi --arch aarch6-linux` doesn't return any URL.. meaning that it could be an issue with any any `aarch64-` platform and this derivation. I'd try to set `platforms = lib.platforms.all` and see what ofborg does with that. In anycase, the PR is generally ready IMO.

**Comment by nim65s** [comment]
I have no idea how meta.platforms could play here, but since you both mention it, I guess we should add it 😅

**Comment by nim65s** [comment]
thanks for the 🪄 !

**Comment by doronbehar** [comment]
@ofborg build casadi

**Comment by doronbehar** [comment]
The aarch64-linux build still appears as failing. I want to make sure it really fails.

**Comment by doronbehar** [comment]
I OK now it is more clear what is the aarch64-linux error:

```
The following tests FAILED:
         66 - ProxQP::dense: test primal infeasibility solving (Failed)
Errors while running CTest
```

@nim65s perhaps could you make an effort to disable that test on `aarch64-linux`?

**Comment by nim65s** [comment]
Done.
There are ~20 tests with the same name, so all are disabled, but I guess this is fine.

**Review by doronbehar — APPROVED** [review]
Changes look good. CI is green now for both Linux architectures, and it was green before the last aarch64-linux commit for x86_64-darwin, and nixpkgs-review was happy as well.

**Review by wegank — APPROVED** [review]
